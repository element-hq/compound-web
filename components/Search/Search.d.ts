import React from "react";
import { Field } from "../Form";
type SearchProps = {
    /**
     * The CSS class name
     */
    className?: string;
    /**
     * The input placeholder.
     * @default "Searchâ€¦"
     */
    placeholder?: string;
    /**
     * The field name.
     */
    name: React.ComponentProps<typeof Field>["name"];
    /**
     * Event handler called when the search changes.
     */
    onChange?: (e: React.ChangeEvent) => void;
};
/**
 * A standalone search component
 */
export declare const Search: ({ className, onChange, placeholder, name, }: SearchProps) => JSX.Element;
export {};
//# sourceMappingURL=Search.d.ts.map