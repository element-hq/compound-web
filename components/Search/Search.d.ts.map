{"version":3,"file":"Search.d.ts","sourceRoot":"","sources":["../../../src/components/Search/Search.tsx"],"names":[],"mappings":"AAiBA,OAAO,KAAK,MAAM,OAAO,CAAC;AAE1B,OAAO,EAAE,KAAK,EAAS,MAAM,SAAS,CAAC;AAKvC,KAAK,WAAW,GAAG;IACjB;;OAEG;IACH,SAAS,CAAC,EAAE,MAAM,CAAC;IACnB;;;OAGG;IACH,WAAW,CAAC,EAAE,MAAM,CAAC;IACrB;;OAEG;IACH,IAAI,EAAE,KAAK,CAAC,cAAc,CAAC,OAAO,KAAK,CAAC,CAAC,MAAM,CAAC,CAAC;IACjD;;OAEG;IACH,QAAQ,CAAC,EAAE,CAAC,CAAC,EAAE,KAAK,CAAC,WAAW,KAAK,IAAI,CAAC;CAC3C,CAAC;AAEF;;GAEG;AACH,eAAO,MAAM,MAAM,gDAMhB,WAAW,KAAG,WAkBhB,CAAC"}