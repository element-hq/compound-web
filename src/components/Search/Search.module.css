/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.search {
  border: 1px solid var(--cpd-color-border-interactive-secondary);
  border-radius: 9999px;
  block-size: 36px;
  box-sizing: border-box;
  color: var(--cpd-color-text-primary);
  display: flex;
  flex-direction: row;
  gap: var(--cpd-space-2x);
  align-items: center;
  padding: var(--cpd-space-1-5x) var(--cpd-space-3x);
}

@media (hover) {
  .search:hover {
    border-color: var(--cpd-color-border-interactive-hovered);
  }
}

.search:active {
  border-color: var(--cpd-color-border-interactive-hovered);
}

.search:focus-within {
  border-color: currentcolor;
}

.icon {
  color: var(--cpd-color-icon-secondary);
  flex-shrink: 0;
}

@media (hover) {
  .search:hover .icon {
    color: var(--cpd-color-icon-primary);
  }
}

.search:active .icon {
  color: var(--cpd-color-icon-primary);
}

.input {
  border: 0;
  background: inherit;
  outline: 0;
  flex: 1;
  min-inline-size: 0;
}

.input::placeholder {
  color: var(--cpd-color-text-placeholder);
}

.input:focus::placeholder {
  color: var(--cpd-color-text-secondary);
}

@media (hover) {
  .search:hover .input::placeholder {
    color: var(--cpd-color-text-secondary);
  }
}
