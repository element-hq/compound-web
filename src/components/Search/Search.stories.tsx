/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { Meta, StoryFn } from "@storybook/react";

import { Search as SearchComponent } from "./Search";
import { Form } from "@radix-ui/react-form";

export default {
  title: "Search",
  component: SearchComponent,
  argTypes: {},
  args: {},
} as Meta<typeof SearchComponent>;

const Template: StoryFn<typeof SearchComponent> = (args) => (
  <Form>
    <SearchComponent {...args} />
  </Form>
);

export const Search = Template.bind({});
Search.args = {};
