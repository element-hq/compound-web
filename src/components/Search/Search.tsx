/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import classnames from "classnames";
import React, { useId } from "react";
import styles from "./Search.module.css";
import { Field, Label } from "../Form";

import SearchIcon from "@vector-im/compound-design-tokens/icons/search.svg";

type SearchProps = {
  /**
   * The CSS class name
   */
  className?: string;
  /**
   * The input placeholder.
   * @default "Search…"
   */
  placeholder?: string;
  /**
   * The field name.
   */
  name: React.ComponentProps<typeof Field>["name"];
  /**
   * Event handler called when the search changes.
   */
  onChange?: (e: React.ChangeEvent) => void;
};

/**
 * A standalone search component
 */
export const Search = ({
  className,
  onChange,
  // TODO: i18n needs to be setup
  placeholder = "Search…",
  name,
}: SearchProps) => {
  const classes = classnames(styles.search, className);
  const id = useId();
  return (
    <Field name={name} asChild>
      <Label className={classes} htmlFor={id}>
        <SearchIcon className={styles.icon} width={20} height={20} />
        <input
          id={id}
          name={name}
          type="search"
          placeholder={placeholder}
          onChange={onChange}
          className={styles.input}
        />
      </Label>
    </Field>
  );
};
