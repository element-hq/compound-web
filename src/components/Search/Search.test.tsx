/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { describe, it, expect, vi } from "vitest";
import { act, render, screen } from "@testing-library/react";
import React, { ChangeEvent } from "react";
import { userEvent } from "@storybook/test";

import { Search } from "./Search";
import { Form } from "@radix-ui/react-form";

type SearchTestProps = {
  onChange?: (e: ChangeEvent) => void;
};

describe("Search", () => {
  const SearchTest = (props: SearchTestProps) => (
    <Form>
      <Search name="search" onChange={props.onChange ?? (() => {})} />
    </Form>
  );

  it("renders", () => {
    const { asFragment } = render(
      <Form>
        <Search name="search" />
      </Form>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("calls onChange when text edited", async () => {
    let value;
    const onChange = vi
      .fn()
      .mockImplementation((e: ChangeEvent<HTMLInputElement>) => {
        value = e.target.value;
      });
    render(<SearchTest onChange={onChange} />);
    const query = "my query";
    const input = screen.getByRole("searchbox");
    await userEvent.type(input, query);

    expect(onChange).toHaveBeenCalled();
    expect(value).toEqual(query);
  });
});
