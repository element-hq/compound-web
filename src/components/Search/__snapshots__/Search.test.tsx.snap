// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Search > renders 1`] = `
<DocumentFragment>
  <form>
    <label
      class="_label_dc87d2 _field_dc87d2 _search_a9d8a5"
      for=":r0:"
    >
      <svg
        class="_icon_a9d8a5"
        fill="none"
        height="20"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M15.0491 16.4633C13.7873 17.4274 12.2105 18 10.5 18C6.35786 18 3 14.6421 3 10.5C3 6.35786 6.35786 3 10.5 3C14.6421 3 18 6.35786 18 10.5C18 12.2105 17.4274 13.7873 16.4633 15.0491L19.7071 18.2929C20.0976 18.6834 20.0976 19.3166 19.7071 19.7071C19.3166 20.0977 18.6834 20.0977 18.2929 19.7071L15.0491 16.4633ZM16 10.5C16 7.46243 13.5376 5 10.5 5C7.46243 5 5 7.46243 5 10.5C5 13.5376 7.46243 16 10.5 16C13.5376 16 16 13.5376 16 10.5Z"
          fill="currentColor"
        />
      </svg>
      <input
        class="_input_a9d8a5"
        id=":r0:"
        name="search"
        placeholder="Searchâ€¦"
        type="search"
      />
    </label>
  </form>
</DocumentFragment>
`;
